### ISSUE TITLE
[type_annotate_public_apis] false positive

### ISSUE THREAD

**Issue by modulovalue** [dart-lang/linter#3268]
Consider the following example:

```dart
abstract class Foo {
  int? get i;
}
class Bar implements Foo {
  @override
  final i = null;
}
```

`Bar.i` violates `type_annotate_public_apis`.

It seems to me that this is a false positive because the type of `Bar.i` is `int?` i.e. it doesn't violate `type_annotate_public_apis ` so the linter should emit no warning for it. 

_(Either that, or `final i = 1;` should also emit a warning?)._

**Comment by srawlins** [comment]
I'll add the extra example in the initial comment:

```
abstract class Foo {
  int? get i;
}
class Bar implements Foo {
  @override
  final i = null; // Lint report
}
class Baz implements Foo {
  @override
  final i = 1; // No lint
}
```

Either lint should be reported on each, or neither, because the type of `Bar.i` and `Baz.i` is inherited.


I think the bug comes from this bizarre definition of "has inferred type":

```
  bool hasInferredType(VariableDeclaration node) {
    var staticType = node.initializer?.staticType;
    return staticType != null &&
        !staticType.isDynamic &&
        !staticType.isDartCoreNull;
  }
```
